// videojs-quality-selector-hls/cjs/ConcreteButton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ConcreteButton;

var _video = _interopRequireDefault(require("video.js"));

var VideoJsButtonClass = _video.default.getComponent('MenuButton');

var VideoJsMenuClass = _video.default.getComponent('Menu');

var VideoJsComponent = _video.default.getComponent('Component');

var Dom = _video.default.dom;
/**
 * Convert string to title case.
 *
 * @param {string} string - the string to convert
 * @return {string} the returned titlecase string
 */

function toTitleCase(string) {
  if (typeof string !== 'string') {
    return string;
  }

  return string.charAt(0).toUpperCase() + string.slice(1);
}
/**
 * Convert string to title case.
 *
 * @param {Player} player - the string to convert
 * @return {MenuButton} the returned titlecase string
 */


function ConcreteButton(player) {
  var _this = this;

  var ConcreteButtonInit = new VideoJsButtonClass(player, {
    title: player.localize('Quality'),
    name: 'QualityButton',
    createItems: function createItems() {
      return [];
    },
    createMenu: function createMenu() {
      var menu = new VideoJsMenuClass(_this.player_, {
        menuButton: _this
      });
      _this.hideThreshold_ = 0; // Add a title list item to the top

      if (_this.options_.title) {
        var titleEl = Dom.createEl('li', {
          className: 'vjs-menu-title',
          innerHTML: toTitleCase(_this.options_.title),
          tabIndex: -1
        });
        var titleComponent = new VideoJsComponent(_this.player_, {
          el: titleEl
        });
        _this.hideThreshold_ += 1;
        menu.addItem(titleComponent);
      }

      _this.items = _this.createItems();

      if (_this.items) {
        // Add menu items to the menu
        for (var i = 0; i < _this.items.length; i++) {
          menu.addItem(_this.items[i]);
        }
      }

      return menu;
    }
  });
  return ConcreteButtonInit;
}

module.exports = exports.default;
